File: Sources/SignalProtocol/messages/SignalMessage.swift
import SignalFfi

class SignalMessage {
    private var handle: OpaquePointer?

    deinit {
        signal_message_destroy(handle)
    }

    internal init(owned rawPtr: OpaquePointer?) {
        handle = rawPtr
    }

    init(bytes: [UInt8]) throws {
        try checkError(signal_message_deserialize(&handle, bytes, bytes.count))
    }

    init(version: UInt8,
         macKey: [UInt8],
         senderRatchetKey: PublicKey,
         counter: UInt32,
         previousCounter: UInt32,
         ciphertext: [UInt8],
         sender senderIdentityKey: PublicKey,
         receiver receiverIdentityKey: PublicKey) throws {
        try checkError(signal_message_new(&handle,
                                          version,
                                          macKey,
                                          macKey.count,
                                          senderRatchetKey.nativeHandle,
                                          counter,
                                          previousCounter,
                                          ciphertext,
                                          ciphertext.count,
                                          senderIdentityKey.nativeHandle,
                                          receiverIdentityKey.nativeHandle))
    }

    func senderRatchetKey() throws -> PublicKey {
        return try invokeFnReturningPublicKey {
            signal_message_get_sender_ratchet_key($0, handle)
        }
    }

    func body() throws -> [UInt8] {
        return try invokeFnReturningArray {
            signal_message_get_body(handle, $0, $1)
        }
    }

    func serialize() throws -> [UInt8] {
        return try invokeFnReturningArray {
            signal_message_get_serialized(handle, $0, $1)
        }
    }

    func messageVersion() throws -> UInt32 {
        return try invokeFnReturningInteger {
            signal_message_get_message_version(handle, $0)
        }
    }

    func counter() throws -> UInt32 {
        return try invokeFnReturningInteger {
            signal_message_get_counter(handle, $0)
        }
    }

    func verifyMac(sender: PublicKey,
                   receiver: PublicKey,
                   macKey: [UInt8]) throws -> Bool {
        var result: Bool = false
        try checkError(signal_message_verify_mac(&result,
                                                 handle,
                                                 sender.nativeHandle,
                                                 receiver.nativeHandle,
                                                 macKey,
                                                 macKey.count))
        return result
    }

    internal var nativeHandle: OpaquePointer? {
        return handle
    }

}

File: java/java/src/main/java/org/whispersystems/libsignal/protocol/SignalMessage.java
/**
 * Copyright (C) 2014-2016 Open Whisper Systems
 *
 * Licensed according to the LICENSE file in this repository.
 */
package org.whispersystems.libsignal.protocol;

import org.signal.client.internal.Native;
import org.whispersystems.libsignal.IdentityKey;
import org.whispersystems.libsignal.InvalidKeyException;
import org.whispersystems.libsignal.InvalidMessageException;
import org.whispersystems.libsignal.LegacyMessageException;
import org.whispersystems.libsignal.ecc.ECPublicKey;
import org.whispersystems.libsignal.util.ByteUtil;

import javax.crypto.spec.SecretKeySpec;

public class SignalMessage implements CiphertextMessage {
  private final long handle;

  @Override
  protected void finalize() {
     Native.SignalMessage_Destroy(this.handle);
  }

  public SignalMessage(byte[] serialized) throws InvalidMessageException, LegacyMessageException {
    handle = Native.SignalMessage_Deserialize(serialized);
  }

  public SignalMessage(long handle) {
    this.handle = handle;
  }

  public SignalMessage(int messageVersion, SecretKeySpec macKey, ECPublicKey senderRatchetKey,
                       int counter, int previousCounter, byte[] ciphertext,
                       IdentityKey senderIdentityKey,
                       IdentityKey receiverIdentityKey)
  {
    handle = Native.SignalMessage_New(messageVersion, macKey.getEncoded(), senderRatchetKey.nativeHandle(),
                 counter, previousCounter, ciphertext,
                 senderIdentityKey.getPublicKey().nativeHandle(),
                 receiverIdentityKey.getPublicKey().nativeHandle());
  }

  public ECPublicKey getSenderRatchetKey()  {
    return new ECPublicKey(Native.SignalMessage_GetSenderRatchetKey(this.handle));
  }

  public int getMessageVersion() {
    return Native.SignalMessage_GetMessageVersion(this.handle);
  }

  public int getCounter() {
    return Native.SignalMessage_GetCounter(this.handle);
  }

  public byte[] getBody() {
    return Native.SignalMessage_GetBody(this.handle);
  }

  public void verifyMac(IdentityKey senderIdentityKey, IdentityKey receiverIdentityKey, SecretKeySpec macKey)
      throws InvalidMessageException
  {
    if(!Native.SignalMessage_VerifyMac(this.handle,
                  senderIdentityKey.getPublicKey().nativeHandle(),
                  receiverIdentityKey.getPublicKey().nativeHandle(),
                  macKey.getEncoded())) {
      throw new InvalidMessageException("Bad Mac!");
    }
  }

  @Override
  public byte[] serialize() {
    return Native.SignalMessage_GetSerialized(this.handle);
  }

  @Override
  public int getType() {
    return CiphertextMessage.WHISPER_TYPE;
  }

  public long nativeHandle() {
    return this.handle;
  }

  public static boolean isLegacy(byte[] message) {
    return message != null && message.length >= 1 &&
        ByteUtil.highBitsToInt(message[0]) != CiphertextMessage.CURRENT_VERSION;
  }

}
